### PR TITLE
Lazy-create thread-specific coverage to immediately process cached edges

### PR DESCRIPTION
**Pull request by evansiroky**
Add one more small improvement that should speed-up calculations during builds that use cached elevations.

### CODE REVIEW

File: src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java [hunk 1]
```diff
@@ -480,31 +509,15 @@ private void assignMissingElevations(
      * Processes a single street edge, creating and assigning the elevation profile.
      *
      * @param ee the street edge
+     * @param edgeGeomerty The geometry of the edge
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java]: typo `edgeGeomerty`

File: src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java [hunk 2]
```diff
@@ -241,7 +241,36 @@ public ProcessEdgeTask(StreetWithElevationEdge swee, ConcurrentHashMap<Long, Cov
...
+                    PackedCoordinateSequence coordinateSequence = cachedElevations.get(
+                        PolylineEncoder.createEncodings(edgeGeometry).getPoints()
+                    );
+                    if (coordinateSequence != null) {
+                        // found a cached value! Set the elevation profile with the pre-calculated data.
+                        setEdgeElevationProfile(swee, coordinateSequence, graph);
+                        edgeNeedsProcessing = false;
+                    }
+                }
+            }
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java]: Extract into its own method.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Please see extract method request.

**Comment by evansiroky** [comment]
Moved some stuff around, hopefully it looks better now.

**Review by binh-dam-ibigroup — DISMISSED** [review]

File: src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java [hunk 3]
```diff
@@ -40,12 +40,16 @@
...
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ForkJoinWorkerThread;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static org.opentripplanner.util.ElevationUtils.computeEllipsoidToGeoidDifference;
 
 /**
+ * THIS CLASS IS MULTI-THEARED
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java]: haha typo. Of course it's practically on the first line 😜

File: src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java [hunk 4]
```diff
@@ -229,36 +238,37 @@ public void buildGraph(Graph graph, GraphBuilderModuleSummary graphBuilderModule
...
+        protected ElevationWorkerThread(ForkJoinPool pool) {
+            super(pool);
         }
 
         /**
-         * Get the thread-specific Coverage instance to avoid multiple threads waiting for a lock to be released on a
-         * Interpolator2D instance.
+         * For unknown reasons, the interpolation of heights at coordinates is a synchronized method in the commonly
+         * used Interpolator2D class. Therefore, it is critical to use a dedicated Coverage instance for each thread to
+         * avoid other threads waiting for a lock to be released on the Coverage instance. This concurrent HashMap will
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java]: There is no longer a Concurrent HashMap, right?

**Review by landonreed — DISMISSED** [review]

**Comment by evansiroky** [comment]
Just added some more code that turns off this multi-threading stuff by default and only activates it when a user sets the `elevationModuleParallelism` in build-config.json to a value greater than 1.

**Review by binh-dam-ibigroup — APPROVED** [review]
Amazing.

**Review by landonreed — APPROVED** [review]

**Comment by landonreed** [comment]
@evansiroky, nice work! I'm going to fix the conflicts and merge.

**Pull request merged**